osu.Game/Skinning/Editor/SkinComponentToolbox.cs

```diff
@@ -57,40 +42,40 @@ private void load()
                 Spacing = new Vector2(2)
             };
 
+            reloadComponents();
+        }
+
+        private void reloadComponents()
+        {
+            fill.Clear();
+
             var skinnableTypes = typeof(OsuGame).Assembly.GetTypes()
                                                 .Where(t => !t.IsInterface)
                                                 .Where(t => typeof(ISkinnableDrawable).IsAssignableFrom(t))
                                                 .OrderBy(t => t.Name)
                                                 .ToArray();
 
             foreach (var type in skinnableTypes)
-            {
-                var component = attemptAddComponent(type);
-
-                if (component != null)
-                {
-                    component.RequestPlacement = t => RequestPlacement?.Invoke(t);
-                    fill.Add(component);
-                }
-            }
+                attemptAddComponent(type);
         }
 
-        private static ToolboxComponentButton attemptAddComponent(Type type)
+        private void attemptAddComponent(Type type)
         {
             try
             {
                 var instance = (Drawable)Activator.CreateInstance(type);
 
                 Debug.Assert(instance != null);
 
-                if (!((ISkinnableDrawable)instance).IsEditable)
-                    return null;
+                if (!((ISkinnableDrawable)instance).IsEditable) return;
 
-                return new ToolboxComponentButton(instance);
+                fill.Add(new ToolboxComponentButton(instance, target)
+                {
+                    RequestPlacement = t => RequestPlacement?.Invoke(t)
+                });
```

Comment:
So if I understand this correctly, this add is what actually holds the invariant that only usable components are shown on their corresponding screens, and it works because the synchronous add causes a synchronous load which fails after dependencies are not found in the borrowing container, and therefore the element is finally not added.

I guess it works but I feel some reservations towards that - it feels a bit *too* magic if I'm honest and needs to be properly xmldoc'd / signposted. Like right now `SongProgress` are visible when skin editing song select because they have nullable dependencies. The try-catch is also doing a lot of magic - any old other failure that *isn't* a DI failure will also make the component disappear. I'd be more fine with catching a DI-specific exception and logging any other unexpected one.

Dunno, probably going to wait for @smoogipoo's thoughts on this.

Reply:
Maybe it's okay for now? Definitely needs documentation though.

Reply:
> The try-catch is also doing a lot of magic - any old other failure that isn't a DI failure will also make the component disappear. I'd be more fine with catching a DI-specific exception and logging any other unexpected one.

I also agree with this bit for better safety, `DependencyNotRegisteredException` should be enough to indicate a DI failure.